libpod/in_memory_state.go

```diff
@@ -303,6 +293,86 @@ func (s *InMemoryState) RemovePod(pod *Pod) error {
 	return nil
 }
 
+// UpdatePod updates a pod's state from the backing database
+// As in-memory states have no database this is a no-op
+func (s *InMemoryState) UpdatePod(pod *Pod) error {
+	return nil
+}
+
+// AddContainerToPod adds a container to the given pod, also adding it to the
+// state
+func (s *InMemoryState) AddContainerToPod(pod *Pod, ctr *Container) error {
+	if !pod.valid {
+		return errors.Wrapf(ErrPodRemoved, "pod %s is not valid and cannot be added to", pod.ID())
+	} else if !ctr.valid {
+		return errors.Wrapf(ErrCtrRemoved, "container %s is not valid and cannot be added to the pod", ctr.ID())
+	}
+
+	if ctr.config.Pod != pod.ID() {
+		return errors.Wrapf(ErrInvalidArg, "container %s is not in pod %s", ctr.ID(), pod.ID())
+	}
+
+	// Add container to pod
+	if err := pod.addContainer(ctr); err != nil {
+		return err
+	}
+
+	// Add container to state
+	_, ok := s.containers[ctr.ID()]
+	if ok {
+		return errors.Wrapf(ErrCtrExists, "container with ID %s already exists in state", ctr.ID())
+	}
+
+	if err := s.ctrNameIndex.Reserve(ctr.Name(), ctr.ID()); err != nil {
+		return errors.Wrapf(err, "error registering container name %s", ctr.Name())
+	}
+
+	if err := s.ctrIDIndex.Add(ctr.ID()); err != nil {
```

Comment:
Should Reserve to the Add internally?

Reply:
These are two packages we inherited from Docker - they use different backing data structures to let them do high-speed lookups on full name (name index) and full or partial ID (ID index). We could modify them to combine them so you don't need to do Add and Reserve, but I'd just as soon leave the code we didn't write alone.

Reply:
Ok. That is fine with me.